### PR TITLE
ARROW-3838: [Rust] CSV Writer

### PR DESCRIPTION
**Pull request by nevi-me**
This supersedes #3111.

Supports:
* writing using the `csv` crate
* customising delimiter, and whether or not to write headers

Limitations:
* writes to `std::fs::File`, I struggled to make it write to any `std::io::Write` interface. @paddyhoran any ideas as you did it for `Reader`?
* values are converted first to `String` before being written. It could be more performant if we could convert directly to byte slice as `csv` crate supports that. I also struggled with this

Potential Further Work:
* writing temporal arrays (after #3726 [ARROW-4386])

### CODE REVIEW

**Comment by nevi-me** [comment]
@andygrove @sunchao @paddyhoran @ntrinquier @liurenjie1024 please review when you get some time.

**Comment by nevi-me** [comment]
AppVeyor failure's because I'm writing to `/tmp/`, which doesn't exist. I'll address that later today

**Review by andygrove — COMMENTED** [review]
@nevi-me This looks great! I think we just need to fix the test to not write to `/tmp`

File: rust/arrow/src/csv/writer.rs [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+        ]);
+        let c3 = PrimitiveArray::<UInt32Type>::from(vec![3, 2, 1]);
+        let c4 = PrimitiveArray::<BooleanType>::from(vec![Some(true), Some(false), None]);
+
+        let batch = RecordBatch::new(
+            Arc::new(schema),
+            vec![Arc::new(c1), Arc::new(c2), Arc::new(c3), Arc::new(c4)],
+        );
+
+        let file = File::create("/tmp/custom_options.csv").unwrap();
```
[the remark this answers is not in the document]
Comment (andygrove) [reply rust/arrow/src/csv/writer.rs]: Yeah .. Appveyor windows build is failing because of this.

File: rust/arrow/src/csv/writer.rs [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+{
+    let c = array.as_any().downcast_ref::<PrimitiveArray<T>>().unwrap();
+    c.value(i).to_string()
+}
+
+/// A CSV writer
+pub struct Writer {
+    /// The file to write to
+    file: File,
+    /// Column delimiter. Defauits to `b','`
```
Comment (ntrinquier) [inline rust/arrow/src/csv/writer.rs]: Small typo here `Defauits` -> `Defaults`

**Review by ntrinquier — COMMENTED** [review]
One minor thing (typo), otherwise LGTM. Thanks!

**Review by paddyhoran — APPROVED** [review]
Other than the comments from others LGTM. A change to use 'Write' can come in a follow up PR. I'm happy to take a look at it as I will need it for the data I have to deal with.

File: rust/arrow/src/csv/reader.rs [hunk 3]
```diff
@@ -395,8 +395,8 @@ pub struct ReaderBuilder {
 }
 
 impl Default for ReaderBuilder {
-    fn default() -> ReaderBuilder {
-        ReaderBuilder {
+    fn default() -> Self {
```
Comment (liurenjie1024) [inline rust/arrow/src/csv/reader.rs]: Maybe it's a better idea to put this refactoring into another PR? It doesn't seem related with csv writer

File: rust/arrow/src/csv/writer.rs [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+    T::Native: ::std::string::ToString,
+{
+    let c = array.as_any().downcast_ref::<PrimitiveArray<T>>().unwrap();
+    c.value(i).to_string()
+}
+
+/// A CSV writer
+pub struct Writer {
+    /// The file to write to
+    file: File,
```
Comment (liurenjie1024) [inline rust/arrow/src/csv/writer.rs]: It may be a better idea to accept a Box<Write> as target. This can make your implementation more flexible.

File: rust/arrow/src/csv/writer.rs [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+    /// Column delimiter. Defauits to `b','`
+    delimiter: u8,
+    /// Whether file should be written with headers. Defaults to `true`
+    has_headers: bool,
```
Comment (liurenjie1024) [inline rust/arrow/src/csv/writer.rs]: Since csv is a text based format, should we also consider encoding?

File: rust/arrow/src/csv/writer.rs [hunk 6]
```diff
@@ -0,0 +1,363 @@
...
+        }
+    }
+
+    /// Write a vector of record batches to a file
+    pub fn write(&self, batches: Vec<&RecordBatch>) -> Result<()> {
+        let mut builder = csv_crate::WriterBuilder::new();
+
+        let mut wtr = builder.delimiter(self.delimiter).from_writer(&self.file);
+
+        if batches.is_empty() {
```
Comment (liurenjie1024) [inline rust/arrow/src/csv/writer.rs]: Check this at beginning of method?

**Review by liurenjie1024 — COMMENTED** [review]
Thanks for the contribution, I left some comments.

File: rust/arrow/src/csv/writer.rs (hunk 5 above)
Comment (nevi-me) [reply rust/arrow/src/csv/writer.rs]: I don't think that the `csv` crate supports different encodings. Only UTF-8 is supported.

File: rust/arrow/src/csv/writer.rs (hunk 6 above)
Comment (nevi-me) [reply rust/arrow/src/csv/writer.rs]: Done, thanks

File: rust/arrow/src/csv/reader.rs (hunk 3 above)
Comment (nevi-me) [reply rust/arrow/src/csv/reader.rs]: The changes are too small, but I can revert them. I wanted to make the reader consistent with the writer

File: rust/arrow/src/csv/writer.rs (hunk 4 above)
Comment (nevi-me) [reply rust/arrow/src/csv/writer.rs]: We can refactor this with a separate PR, that ideally does the same to the CSV reader, so we're consistent with our IO APIs

File: rust/arrow/src/csv/writer.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (nevi-me) [reply rust/arrow/src/csv/writer.rs]: @andygrove I've changed it to `target/..`, I opted not to use any of the folders inside target as they aren't always guaranteed to exist from a clean build

File: rust/arrow/src/csv/writer.rs (hunk 4 above)
Comment (liurenjie1024) [reply rust/arrow/src/csv/writer.rs]: Great, looking forward to see the following PR

File: rust/arrow/src/csv/writer.rs (hunk 5 above)
Comment (liurenjie1024) [reply rust/arrow/src/csv/writer.rs]: ok

**Comment by liurenjie1024** [comment]
LGTM

**Comment by nevi-me** [comment]
@andygrove @paddyhoran I've fixed the tmp file issue
